File: resolvedomains/resolvedomains.py
#!/usr/bin/env python3
import sys
import re
import concurrent.futures
import dns.resolver

class Resolver(object):
    def __init__(self, string):
        self.ip_addresses = []
        if Resolver._is_ip_addr(string):
            self.ip_addresses.append(string)
        else:
            resolved = Resolver._resolve(string)
            if not resolved is None:
                self.ip_addresses += resolved

    def is_domain_here(self, domain):
        ip_addresses = Resolver._resolve(domain)
        if ip_addresses is None: return False
        for ip_addr in ip_addresses:
            if ip_addr in self.ip_addresses:
                return True
        return False

    @staticmethod
    def _is_ip_addr(string):
        return re.match('''^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$''', string) is not None

    @staticmethod
    def _resolve(domain):
        try:
            return [str(ip_addr) for ip_addr in dns.resolver.query(domain, 'A')]
        except (dns.resolver.NXDOMAIN, dns.resolver.NoAnswer, dns.resolver.NoNameservers, dns.name.EmptyLabel) as e:
            return None

def solve_ip_addresses(s):
    r = Resolver(s)
    return r.ip_addresses

def main():
    lines = list(set(sys.stdin.read().splitlines()))
    resolved = []
    with concurrent.futures.ThreadPoolExecutor(max_workers=30) as executor:
        future_addresses = {executor.submit(solve_ip_addresses, line): line for line in lines}
        for f in concurrent.futures.as_completed(future_addresses):
            addresses = f.result()
            for ip_addr in addresses:
                if not ip_addr in resolved:
                    resolved.append(ip_addr)
                    print(ip_addr)
                
if __name__ == '__main__':
    main()

File: resolvedomains/resolvedomains.rb
require 'resolv'
require 'set'

class Resolver
  attr_reader :ip_addresses

  def initialize(string)
    @ip_addresses = []
    if Resolver.is_ip_addr?(string)
      @ip_addresses << string
    else
      resolved = Resolver.resolve(string)
      @ip_addresses += resolved unless resolved.nil?
    end
  end

  def is_domain_here?(domain)
    ip_addresses = Resolver.resolve(domain)
    return false if ip_addresses.nil?
    ip_addresses.any? { |ip_addr| @ip_addresses.include?(ip_addr) }
  end

  def self.is_ip_addr?(string)
    !!(string =~ /^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/)
  end

  def self.resolve(domain)
    begin
      Resolv.getaddresses(domain).map(&:to_s)
    rescue Resolv::ResolvError, Resolv::NoAnswer, Resolv::NoNameservers => e
      nil
    end
  end
end

def solve_ip_addresses(s)
  r = Resolver.new(s)
  r.ip_addresses
end

def main
  lines = $stdin.read.split("\n").uniq
  resolved = Set.new
  lines.each_slice(30) do |batch|
    batch_resolved = batch.map { |line| solve_ip_addresses(line) }.flatten.uniq
    batch_resolved.each do |ip_addr|
      unless resolved.include?(ip_addr)
        resolved << ip_addr
        puts ip_addr
      end
    end
  end
end

main
